### PR TITLE
[Serve] Make `max_batch_size` and `batch_wait_timeout_s` reconfigurable

### LINKED ISSUE #36844: [Serve] Let users modify `max_batch_size` and `batch_wait_timeout_s` in `@serve.batch` dynamically
### Description

Users can set `max_batch_size` and `batch_wait_timeout_s` in the `@serve.batch`, but there's no way to modify these parameters dynamically during runtime or through a config file. Users should be able to modify the batch size and timeout using setter methods on the batch handler:

```python
...

@serve.batch(max_batch_size=5, batch_wait_timeout_s=30)
def batch_handler(self, request_list):
    ...

def modify_batch_params(self):
    """Users should be able to modify these parameters using setter methods."""
    self.batch_handler.set_max_batch_size(10)
    self.batch_handler.set_batch_wait_timeout_s(20)

...
```

### Use case

This features lets users manually reconfigure their batching parameters during runtime through the `reconfigure` method. Advanced users can also dynamically tune their batch size or timeout based on the `batch_handler`'s current workload.

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

The `@serve.batch` decorator has two parameters: `max_batch_size` and `batch_wait_timeout_s`. These parameters can be set in the decorator. However, they cannot be reconfigured after the Serve application starts.

This change adds two setter methods: `set_max_batch_size` and `set_batch_wait_timeout_s`. Users can reconfigure their `@serve.batch` parameters using these methods:

```python
@serve.batch(max_batch_size=1, batch_wait_timeout_s=0.1)
def batch_handler(self, request_list):
    ...

self.batch_handler.set_max_batch_size(5)
self.batch_handler.set_batch_wait_timeout_s(0.5)
```

## Related issue number

Closes #36844

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - This change adds unit tests to `test_batching.py`.

### CODE REVIEW

**Comment by shrekris-anyscale** [comment]
@edoakes the code changes are ready for review. I'm planning on adding doc changes to this PR soon.

**Comment by shrekris-anyscale** [comment]
@edoakes I've added documentation. This PR is ready for review end-to-end.

File: python/ray/serve/batching.py [hunk 1]
```diff
@@ -114,11 +116,15 @@ async def wait_for_batch(self) -> List[Any]:
         batch = []
         batch.append(await self.queue.get())
 
+        # Cache current max_batch_size and batch_wait_timeout_s for this batch
+        max_batch_size = self.max_batch_size
+        batch_wait_timeout_s = self.batch_wait_timeout_s
+
```
Comment (edoakes) [inline python/ray/serve/batching.py:122]:
```suggestion
        # Cache current max_batch_size and batch_wait_timeout_s for this batch.
        max_batch_size = self.max_batch_size
        batch_wait_timeout_s = self.batch_wait_timeout_s

```

Comment (edoakes) [reply python/ray/serve/batching.py:122]: so the semantics here are that the settings will be picked up after the current batch finishes? sounds fine to me

**Review by edoakes — APPROVED** [review]

**Review by edoakes — COMMENTED** [review]
@shrekris-anyscale can you add a sentence to the `@serve.batch` docstring mentioning the setters?

File: python/ray/serve/batching.py (hunk 1 above)
Comment (shrekris-anyscale) [reply python/ray/serve/batching.py]: Yep, that's right.

**Comment by shrekris-anyscale** [comment]
@edoakes Thanks, I updated the docstring. Please take a look.

**Review by angelinalg — APPROVED** [review]

**Pull request merged**